### DISCUSSION TITLE
"bv_maker build" ... failed

### DISCUSSION THREAD

**Discussion by francoisLeroy** [brainvisa/brainvisa.github.io#74] [Q&A] [answered]
Hello,
I tried to update the Brainvisa development package in my casa_distro v3.1 environment, because of new git migration.
I ran twice "bv_maker sources" and "bv_maker configure" --> all was ok the second time
but "bv_maker build --clean" failed. I also tried "bv_maker build" without cleaning old files but it also failed.
Please find attached the log file.  Thanks for your help!
Francois
[bv_maker_failed.txt](https://github.com/brainvisa/brainvisa.github.io/files/7215674/bv_maker_failed.txt)

**Comment by denisri** [comment]
The sources don't seem updated actually, the error you get is in `cortical_surface-private/bug_fix`, where it should now be named `cortical_surface-nonfree/master`. The update takes several steps because it needs to:
1. get the new sources of `bv_maker` with the new URLs of the projects on GitHub (done with a first `bv_maker sources`)
2. install this new version of bv_maker (done with `bv_maker configure`)
3. then re-update sources to actually get the new sources (done with another run of `bv_maker sources`)
4. configure and build the new sources (dons with `bv_maker configure build`)

Normally if I'm not wrong the whole process should work using twice a full `bv_maker` command (without arguments, which should do `sources`, `configure`, `build` steps even when it's too much). Here if you have done twice the `sources` step without `configure` between the two `sources`, then it is not enough.
Anyway at the point you are, running another `bv_maker` should do the job.

**Comment by francoisLeroy** [comment]
Hi,
I tried running bv_maker twice but I have still some errors.

[bv_log_2.txt](https://github.com/brainvisa/brainvisa.github.io/files/7244657/bv_log_2.txt)

Here is the log file.
Thanks for your help

**Comment by denisri** [comment]
It has still not updated the sources.
Can we have a look at your `/casa/host/conf/bv_maker.cfg` ? Are there any customizations there ?

**Comment by francoisLeroy** [comment]
Sure. Here is the bv_maker.cfg
I guess I need to do something about cortical-surface-private but I don't know exactly what.
Thanks again
[bv_maker.txt](https://github.com/brainvisa/brainvisa.github.io/files/7245202/bv_maker.txt)

**Comment by denisri** [comment]
AAAAAAh, that's it !
You have specified manually the `cortical_surface` projects, with their address and location, and their location has changed...
You can avoid specifying the full URL for projects already known by bv_maker, either using a distro / projects group which contain the ones you need (`bvdev` for instance), or using a config file like this (but you still need to update the projects names when they have changed):

```
[ source $CASA_SRC ]
  brainvisa opensource bug_fix
  brainvisa cortical_surface-nonfree bug_fix
  brainvisa morphologist-nonfree bug_fix
  + perso/fleroy perso/fleroy/bug_fix
  + baby bug_fix


[ build $CASA_BUILD ]
  build_type = Release
  make_options = -j8
  opensource bug_fix $CASA_SRC
  brainvisa cortical_surface-nonfree bug_fix $CASA_SRC
  brainvisa morphologist-nonfree bug_fix $CASA_SRC
  brainvisa baby bug_fix $CASA_SRC
  + $CASA_SRC/perso/fleroy/bug_fix
```

**Reply by francoisLeroy** [reply]
Thanks a lot for this update! bv_maker ran much further this time
However, I have issues related to my old c++ code in my perso/fleroy repository. for instance, I have this line (in a .h file)
#include <cortical_surface/mesh/isoLine.h> which produces an error.
How should I change it according to the new cortical_surface-nonfree package?
Thanks
Francois

**Comment by denisri** [comment]
Hmm, the contents have not changed at all, just the projects have been renamed and moved to another server. The include you cite should still be OK. Maybe the symlinks already in `/casa/build/include` and pointing to the old sources location are not redone.
I think you should do (in the container) a:
```
bv_clean_build_tree /casa/host/build
```
then retry `bv_maker build`

**Comment by francoisLeroy** [comment]
Hi,
It seems to improve because I get another error at the link stage this time
ie "cannot find -lcortical_surface"
Please have a look at the log file.
Thanks!
[bv_log_3.txt](https://github.com/brainvisa/brainvisa.github.io/files/7276711/bv_log_3.txt)

**Comment by denisri** [comment]
Have you adapted the CMakeLists.txt of your personal project ? I see in the svn repo:
```
BRAINVISA_FIND_PACKAGE( cortical_surface-private REQUIRED )
```
which should be changed to:
```
BRAINVISA_FIND_PACKAGE( cortical_surface-nonfree REQUIRED )
```
Now I'm not sure it is the cause of the problem.
After the build, do you have the file
`/casa/host/build/lib/libcortical_surface.so` ?

**Comment by francoisLeroy** [comment]
I tried 
BRAINVISA_FIND_PACKAGE( cortical_surface-nonfree REQUIRED )
but it did not work out.
Then 
BRAINVISA_FIND_PACKAGE( cortical_surface-gpl REQUIRED )
seems better because I have no error at the "configure" stage
but I have still the same issue at the "build" stage ie "cannot find -lcortical_surface"
It is bizarre because there is libcortical_surface.so in /casa/host/build/lib
It is a link towards libcortical_surface.so.5.1.0 which also exists in /casa/host/build/lib
Puzzling, no?
Thanks for your help

**Comment by denisri** [comment]
> but it did not work out.

You should always precise what's the result, the error message, because like that it doesn't help to understand ;)

> BRAINVISA_FIND_PACKAGE( cortical_surface-gpl REQUIRED )
> seems better because I have no error at the "configure" stage

but it's not the right dependency and doesn't allow to link against the library

> It is bizarre because there is libcortical_surface.so in /casa/host/build/lib
> It is a link towards libcortical_surface.so.5.1.0 which also exists in /casa/host/build/lib

Yes it is.
What's the error using cortical_surface-nonfree ? I think the problem is rather there...

**Comment by francoisLeroy** [comment]
Here it is. Thanks
[bv_configure.txt](https://github.com/brainvisa/brainvisa.github.io/files/7277866/bv_configure.txt)

**Comment by denisri** [comment]
Strange. It doesn't seem to find the project `cortical_surface-nonfree`, whereas it is actually here since we have already fixed your config for it, and you have its library built in `build/lib`.
I tried getting your sources in `perso/fleroy/bug_fix`, and after editing the `CMakeLists.txt` as said above, for me it worked, I have the command `AimsLabels2Tex_perso` compiled. So I don't understand.
Does the command:
```
cd /casa/host/build; make cortical_surface
```
still work ?

**Comment by francoisLeroy** [comment]
Here it is:
fl985244@is151617:~$ bv bash                                                                     (opensource-master-5.0)
fl985244@is151617:/home/fl985244$ cd /casa/host/build/                    (opensource-master-5.0)
fl985244@is151617:/casa/host/build$ make cortical_surface                 
make: *** No rule to make target 'cortical_surface'.  Stop.                                      (opensource-master-5.0)

**Comment by denisri** [comment]
So that's it: `cortical_surface-nonfree` is not built. It has been done at some time because you still have the library file, but it is not handled any longer in the build system. Can we check your `bv_maker.cfg` again please ?

**Comment by francoisLeroy** [comment]
[ source $CASA_SRC ]                                                                               
brainvisa opensource bug_fix                                                                     
+ brainvisa cortical_surface-nonfree bug_fix                                                     
+ brainvisa morphologist-nonfree bug_fix                                                         
+ baby bug_fix                                                                                   
+ perso/fleroy perso/fleroy/bug_fix      
                                                                                                                                                                                                                                                        [ build $CASA_BUILD ]                                                                              
build_type = Release                                                                             
make_options = -j8                                                                               
opensource bug_fix $CASA_SRC                                                                     
brainvisa cortical_surface-nonfree bug_fix $CASA_SRC                                             
brainvisa morphologist-nonfree bug_fix $CASA_SRC                                                 
brainvisa baby bug_fix $CASA_SRC                                                                 
+ $CASA_SRC/perso/fleroy/bug_fix

**Comment by c-langlet** [comment]
Hi,

Can you try renaming "bug_fix" into "master", bug_fix is the name of branches used by subversion while master is the one used by git. I don't know about your personal repositories though.

If this does solve the issue, you should think about replacing it by $CASA_BRANCH and making sure in your casa_distro.json that you have "branch": "master".

Keep us informed.

**Comment by denisri** [comment, marked as answer]
@c-langlet  is right, but I guess it's not the key of the proiblem since we still maintain "`bug_fix`" as an alias to '"`master`"; however it's good to change it for the future.
The `bv_maker.cfg` you print here is difficult to read, it's not the raw text. You should use triple inverted quotes in the github editor around the raw text (or use an attached file as you used to do before). But I have the _impression_ that you have used both a `+` sign and "`brainvisa`" in your sources section, which is redundant and confusing for `bv_maker` (and was not in the model I have posted earlier), and may cause it to ignore the source lines (silently, which is not good I admit). It is either `+` or `brainvisa`, not both. Your bv_maker.cfg should look like this (exactly, every sign is important: no `+ brainvisa something`):
```
[ source $CASA_SRC ]
  brainvisa opensource $CASA_BRANCH
  brainvisa cortical_surface-nonfree $CASA_BRANCH
  brainvisa morphologist-nonfree $CASA_BRANCH
  brainvisa baby $CASA_BRANCH
  + perso/fleroy perso/fleroy/bug_fix

[ build $CASA_BUILD ]
  build_type = Release
  make_options = -j$NCPU
  opensource $CASA_BRANCH $CASA_SRC
  brainvisa cortical_surface-nonfree $CASA_BRANCH $CASA_SRC
  brainvisa morphologist-nonfree $CASA_BRANCH $CASA_SRC
  brainvisa baby $CASA_BRANCH $CASA_SRC
  + $CASA_SRC/perso/fleroy/bug_fix
```
By the way your personal project seems to contain a duplicate version of itself:
```
ls /casa/host/src/perso/fleroy/bug_fix
AimsLabels2Tex_perso      AimsSulcusDepthProfile  CMakeLists.txt     project_info.cmake
AimsSulcalDepthMapping    AimsSulcusT2Mapping     bbsHomotopicSnake  src.pro
AimsSulcalTextureMapping  AimsTex2Cx              bug_fix

ls /casa/host/src/perso/fleroy/bug_fix/bug_fix/
AimsLabels2Tex_perso      AimsSulcusDepthProfile  CMakeLists.txt      src.pro
AimsSulcalDepthMapping    AimsSulcusT2Mapping     bbsHomotopicSnake
AimsSulcalTextureMapping  AimsTex2Cx              project_info.cmake
```

**Comment by francoisLeroy** [comment]
Great, it is successful! Thanks a lot
I just had to change $CASA_SRC/perso/fleroy/bug_fix
                        into $CASA_SRC/perso/fleroy/$CASA_BRANCH 
in both "Source" and "Build" sections, and then it was all fine.
By the way, I removed the duplicate "bug_fix" directory you mentioned in my local repository.
Is it easy to remove it on the server repository?

**Comment by denisri** [comment]
The problem was rather in the declaration of `cortical_surface-nonfree` and `morphologist-nonfree`; anyway it's fine now that it works ;)
To remove the duplicate on the server I guess you just have to do:
```
svn rm /casa/host/src/perso/fleroy/bug_fix/bug_fix
svn ci /casa/host/src/perso/fleroy/bug_fix
```

**Comment by francoisLeroy** [comment]
It worked, thanks!
